### ISSUE TITLE
Missing ids on windows

### ISSUE THREAD

**Issue by jeremythille** [spencermountain/dumpster-dive#77]
Trying to parse the big 65.7 GB English wiki, enwiki-20190101-pages-articles-multistream.xml.

Parsing goes well, data gets saved to Mongo, but at some point, every line is the same.

> current: 471,877 pages - "Shangluo"
> current: 471,877 pages - "Shangluo"
> current: 471,877 pages - "Shangluo"
> current: 471,877 pages - "Shangluo"

about 200 times, and nothing else after that.

If I CTRL + C to break the process, then I suddenly get 8 times this error (one per worker, I guess)

> one sec, cleaning-up the workers...
> --uncaught top-process error--

> { ProcessTerminatedError: cancel after 0 retries!
>     at tasks.filter.forEach.task (C:\Users\Jeremy\AppData\Roaming\npm\node_modules\dumpster-dive\node_modules\worker-nodes\lib\pool.js:111:39)
>     at Array.forEach (<anonymous>)
>     at WorkerNodes.handleWorkerExit (C:\Users\Jeremy\AppData\Roaming\npm\node_modules\dumpster-dive\node_modules\worker-nodes\lib\pool.js:110:14)
>     at Worker.worker.on.exitCode (C:\Users\Jeremy\AppData\Roaming\npm\node_modules\dumpster-dive\node_modules\worker-nodes\lib\pool.js:160:44)
>     at Worker.emit (events.js:182:13)
>     at WorkerProcess.Worker.process.once.code (C:\Users\Jeremy\AppData\Roaming\npm\node_modules\dumpster-dive\node_modules\worker-nodes\lib\worker.js:39:18)
>     at Object.onceWrapper (events.js:273:13)
>     at WorkerProcess.emit (events.js:182:13)
>     at ChildProcess.WorkerProcess.child.once.code (C:\Users\Jeremy\AppData\Roaming\npm\node_modules\dumpster-dive\node_modules\worker-nodes\lib\worker\process.js:42:41)
>     at Object.onceWrapper (events.js:273:13)
>   name: 'ProcessTerminatedError',
>   message: 'cancel after 0 retries!' }

I launched the parser 4 times, it skips what was done previously, then gets to this again :

> current: 471,877 pages - "Shangluo"
> current: 471,877 pages - "Shangluo"
> current: 471,877 pages - "Shangluo"
> current: 471,877 pages - "Shangluo"

same error, and stops here every time.

**Comment by spencermountain** [comment]
hi Jeremy, yikes. I've never seen that behavior before.
I tried Shangluo with the parser, and don't see any funny-business.
i've just updated the deps in `5.1.2`, i doubt that will fix anything though. Have you tried using newer node versions? `worker-nodes` may be sensitive to that.

otherwise, it may be an issue for [their library](https://github.com/allegro/node-worker-nodes).
let me know how it goes
cheers

**Comment by jeremythille** [comment]
Thanks for the reply, sorry for not responding earlier, I was on holiday.

I have updated Node from 10 to 11. The behaviour is still the same : the parser stops after a while. The recurring line has changed, though. It is now :

>  current: 1 pages - "true"
>  current: 1 pages - "true"
>  current: 1 pages - "true"
>  current: 1 pages - "true"
>  current: 1 pages - "true"

I get this about 200 times,  then, nothing. It seems to be idle, CPU usage 0%.

And if I hit Ctrl + C to stop the process, I get 8 times this error:

> one sec, cleaning-up the workers...
> --uncaught top-process error--
> { ProcessTerminatedError: cancel after 0 retries!
>     at tasks.filter.forEach.task (C:\Users\Jeremy\AppData\Roaming\npm\node_modules\dumpster-dive\node_modules\worker-nodes\lib\pool.js:111:39)
>     at Array.forEach (<anonymous>)
>     at WorkerNodes.handleWorkerExit (C:\Users\Jeremy\AppData\Roaming\npm\node_modules\dumpster-dive\node_modules\worker-nodes\lib\pool.js:110:14)
>     at Worker.worker.on.exitCode (C:\Users\Jeremy\AppData\Roaming\npm\node_modules\dumpster-dive\node_modules\worker-nodes\lib\pool.js:160:44)
>     at Worker.emit (events.js:193:13)
>     at WorkerProcess.Worker.process.once.code (C:\Users\Jeremy\AppData\Roaming\npm\node_modules\dumpster-dive\node_modules\worker-nodes\lib\worker.js:39:18)
>     at Object.onceWrapper (events.js:281:20)
>     at WorkerProcess.emit (events.js:193:13)
>     at ChildProcess.WorkerProcess.child.once.code (C:\Users\Jeremy\AppData\Roaming\npm\node_modules\dumpster-dive\node_modules\worker-nodes\lib\worker\process.js:42:41)
>     at Object.onceWrapper (events.js:281:20)
>   name: 'ProcessTerminatedError',
>   message: 'cancel after 0 retries!' }

The executed command is :

> dumpster enwiki-20190101-pages-articles-multistream.xml

**Comment by spencermountain** [comment]
yeah, you need to find out what it's writing to the database.
my bet is it's all nulls or, id:null or something like that

**Comment by spencermountain** [comment]
*what it's trying to write to the database

**Comment by jeremythille** [comment]
No, actually now it writes one document (!) then stops. Before, I got thousands of documents written to Mongo before I got the Shangluo problem.

Documents written to Mongo look like this :

```
{
    "_id" : "",
    "title" : true,
    "pageID" : true,
    "categories" : [ 
        "Agronomy", 
        "Environmental science software"
    ],
    "sections" : [ 
        {
            "title" : "",
            "depth" : 0,
            "paragraphs" : [ 
                {
                    "sentences" : [ 
                        {
                            "text" : "The Decision Support System for Agrotechnology Transfer (DSSAT) is a set of computer programs for simulating agricultural crop growth.",
                            "links" : [ 
                                {
                                    "page" : "Decision support system",
                                    "text" : "Decision Support System"
                                }
                            ],
                            "formatting" : {
                                "bold" : [ 
                                    "Decision Support System for Agrotechnology Transfer", 
                                    "DSSAT"
                                ]
                            }
                        }, 
                        {
                            "text" : "It has been used in over 100 countries by agronomists for evaluating farming methods.",
                            "links" : [ 
                                {
                                    "page" : "Agronomy",
                                    "text" : "agronomists"
                                }
                            ]
                        }, 
                        {
                            "text" : "One application has been assessing the possible impacts on agriculture of climate change and testing adaptation methods."
                        }
                    ]
                }, 
```

etc. Truncated the rest because it's big

**Comment by spencermountain** [comment]
yeah, `id:""` - that's the problem. are you creating this json yourself? you can see how this library creates the `_id` property [here](https://github.com/spencermountain/dumpster-dive/blob/master/src/worker/02-parseWiki.js#L48)

**Comment by jeremythille** [comment]
No I copied it from Robo3T. This data has been written to Mongo, in an empty collection, when I ran 

> dumpster enwiki-20190101-pages-articles-multistream.xml

I didn't create it myself, I just copied it.

**Comment by jeremythille** [comment]
I have recorded and shared a short video of what happens when I launch Dumpster from scratch, with an empty database.

https://drive.google.com/file/d/150tSIacCjq8WlmBBtjnBUtABSHJjHUZZ/view?usp=sharing

Only one record is created in Mongo, with an empty _id. Since all records seem to have an empty _id, all following records are considered duplicates and skipped.

Node version : 11.14.0
Mongo version : 3.4.6

**Comment by spencermountain** [comment]
thanks Jeremy, thought of what may be happening this morning, will try to release a fix this afternoon

**Comment by spencermountain** [comment]
in the mean-time, you can try something like this:
```js
const dumpster=require(''dumpster-dive)
let obj={
	file: path,
	db: dbName,
	custom: function(doc) {
               let data=doc.json()
              delete data._id
	      return data
	}
}
dumpster(obj, () => console.log('done') )
```

**Comment by jeremythille** [comment]
Thanks, I tried your code above, with the Afrikaans XML file. Several records get saved correctly, but I almost immediately get this error from all workers :

```
         ---------------------------
                  oh hi �
         size:             386.8 MB
                          8 workers
                       48.4 MB each
         estimate:         1.1 mins
         ---------------------------


 #4  +200 pages  -  1s - "Wit Nyl"
 #1  +200 pages  -  1s - "Thomas Boydell"
   ====DB write error (worker 3)===
Error: key Skaal (lj.) must not contain '.'
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:913:19)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:727:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:727:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:937:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:727:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)

====DB write error (worker 6)===
Error: key nr. must not contain '.'
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:913:19)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:727:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:727:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:937:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:727:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
```

I think I'll just wait for the fix :)

**Comment by spencermountain** [comment]
ah, this was my bad. 
Fixed in `5.2.0`.
thanks for your help!

**Comment by jeremythille** [comment]
Installed 5.2.0 and still getting the same error :/

```
$ npm i dumpster-dive

+ dumpster-dive@5.2.0
added 13 packages from 7 contributors and audited 329 packages in 3.41s
found 0 vulnerabilities
```

Then I get the following output :

```
         ---------------------------
                  oh hi �
         size:             386.8 MB
                          8 workers
                       48.4 MB each
         estimate:         1.1 mins
         ---------------------------


-- 200  duplicate pages --
 #4  +0 pages  -  1s - "Wit Nyl"
-- 200  duplicate pages --
 #1  +0 pages  -  1s - "Thomas Boydell"
   ====DB write error (worker 3)===
Error: key Skaal (lj.) must not contain '.'
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:913:19)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:727:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:727:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:937:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
    at serializeInto (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:727:17)
    at serializeObject (J:\Wikipedia\node_modules\bson\lib\bson\parser\serializer.js:347:18)
```

**Comment by spencermountain** [comment]
is that from the custom function, or the default one-liner?

**Comment by jeremythille** [comment]
It was using the custom function, but I just tried with the default one-liner, and I get the same error (`Error: key red. must not contain '.'`)

**Comment by spencermountain** [comment]
Weird. I’m not sure what’s going wrong. For what it’s worth, that key error
should only fail that record, i think. The process should still work just
fine.

Do the tests pass?

`npm run test`


I can take a look at the `Skaal (lj.)` page, but that shouldn’t be related
to the issue you had yesterday

cheers


On April 25, 2019 at 10:34:04 AM, Jeremy Thille (notifications@github.com)
wrote:

It was using the custom function, but I just tried with the default
one-liner, and I get the same error (Error: key red. must not contain '.')

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub
<https://github.com/spencermountain/dumpster-dive/issues/77#issuecomment-486698363>,
or mute the thread

**Comment by jeremythille** [comment]
> that key error should only fail that record, i think. The process should still work just fine.

Oh yes, since 5.2.0 it works again, it only errors and skips titles with a `.` in them. The other records are processed and saved like before (before I updated Node from 10 to 11). Sorry, I forgot to mention this, and I focused on the errors!

> Do the tests pass? `npm run test`

No, the test fails, here's the full output :

```
$ npm run test

> dumpster-dive@5.2.0 test D:\Temp\dumpster-dive-master
> node ./scripts/test.js

running cli tests:

  test-real-smallwiki

'.' is not recognized as an internal or external command,
operable program or batch file.

    × count-is-approx
    ------------------
      operator: ok
      expected: true
      actual:   false
      at: db.count.count (D:\Temp\dumpster-dive-master\tests\cli.test.js:11:9)
      stack: |-
  Error: count-is-approx
  at Test.assert [as _assert] (D:\Temp\dumpster-dive-master\node_modules\tape\lib\test.js:226:54)
  at Test.bound [as _assert] (D:\Temp\dumpster-dive-master\node_modules\tape\lib\test.js:77:32)
  at Test.assert (D:\Temp\dumpster-dive-master\node_modules\tape\lib\test.js:344:10)
  at Test.bound [as ok] (D:\Temp\dumpster-dive-master\node_modules\tape\lib\test.js:77:32)
  at db.count.count (D:\Temp\dumpster-dive-master\tests\cli.test.js:11:9)
  at col.countDocuments.then.len (D:\Temp\dumpster-dive-master\tests\db.js:27:7)
  at processTicksAndRejections (internal/process/task_queues.js:86:5)

  Failed Tests: There was 1 failure
    test-real-smallwiki
      × count-is-approx

  total:     1
  passing:   0
  failing:   1
  duration:  7.4s

running custom tests:
  custom-made-tinywiki

             ---------------------------
                      oh hi �
             size:              26.3 kB
                              2 workers
                           13.1 kB each
             estimate:         0.0 mins
             ---------------------------
        �  worker #1 has finished �
    -- 1  duplicate pages --
     #0  +7 pages  -  1s - "Hello"
        �  worker #0 has finished �
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
         current: 7 pages - "Bodmin"
```

Even though the parser works (except for titles with a `.` in them), it always bumps and gets stuck on one particular entry sooner or later, and never finishes.

With the full English Wiki, it was `471,877 pages - "Shangluo"`.
With the Afrikaans wiki, it is `101,297 pages - "Hereford (dubbelsinnig)"`
Even the test seem to bump and get stuck at `7 pages - "Bodmin"`

**Comment by spencermountain** [comment]
weird!!

Yeah, this has got to be a windows thing. I've tested this on mac/linux.

you know what,  can you do me a favour and [checkout the parser](https://github.com/spencermountain/wtf_wikipedia/) and make sure those tests pass? (`npm run test`).
I don't have a windows computer strong-enough to run this program on.

There's a lot of regexes in there, and one may be using a sloppy newline. If any of those tests fail, then we know the problem. If it passes, then there's only [4 or 5](https://github.com/spencermountain/dumpster-dive/blob/master/src/worker/01-parsePage.js) regexes in this library to look at.

cheers!

**Comment by jeremythille** [comment]
The test for wtf_wikipedia fails to run because :

```
Jeremy@JEREMY-PC D:\Temp\wtf_wikipedia-master
$ npm run test

> wtf_wikipedia@7.3.0 test D:\Temp\wtf_wikipedia-master
> node ./scripts/test.js

'.' is not recognized as an internal or external command,
operable program or batch file.
```
and it just ends here.

I tried removing all the `./` from package.json, same thing.

**Comment by spencermountain** [comment]
ah, crap.
how about this:
```bash
npm i -g tape tap-spec
tape  tests/*.test.js | tap-spec
```

**Comment by jeremythille** [comment]
```
npm i -g tape tap-spec
tape  tests\*.test.js | tap-spec
```
worked perfectly, all tests are passing :
```
 total:     3178
  passing:   3178
  duration:  5.8s
```
I tried the same command for Dumpster-dive, but I got the same output as before.

> sorry for the fuss!

Don't worry, we all know how it is :)

**Comment by spencermountain** [comment]
ah, thank you. that's great.
I'm certain this is a windows issue. You may be the first to use this lib in windows. Thank you for your patience.

it looks like the page content is getting parsed properly, but the title and id sections are getting missed. Those regexes are [here](https://github.com/spencermountain/dumpster-dive/blob/master/src/worker/01-parsePage.js#L25).

don't have time today, but will poke-around at those two regexes asap
cheers

**Comment by spencermountain** [comment]
hey, changed a regex that may be newline sensitive, wanna try `5.2.1`
thanks

**Comment by jeremythille** [comment]
Upgraded to 5.2.1, to no avail :/

Running `node index.js` (the workaround you gave me the other day) produces an avalanche of errors.

Running `dumpster afwiki-latest-pages-articles.xml` produces the same thing. After a while, it still gets stuck at `101,297 pages - "Hereford (dubbelsinnig)"`, which repeats again and again and again.

I've recorded my terminal again, if it helps ( I can't copy the full output here, there's way too much of it) : https://drive.google.com/file/d/17gRvNulXVy7eFHcqow6BXKWECHSrVfgy/view?usp=sharing

**Comment by spencermountain** [comment]
thanks for the great report.

yeah, the ids are still not getting through for some reason. The page titles look fine.

yeah, that `delete data._id` thing shouldn't be necessary anymore. 
actually, I mean for now, you can try setting data._id as [a random string](https://stackoverflow.com/questions/1349404/generate-random-string-characters-in-javascript) each time. (with no `.`!) 

the `._id` param does have some [funny business here](https://github.com/spencermountain/dumpster-dive/blob/master/src/worker/02-parseWiki.js#L52). A good `console.log` there would print the ids,
My guess is somehow one of these values is `true`, and all the `._ids` are being set to true.

I can try another release to test this, if you'd like.

We should be able to find a solution. Also, you can try it in docker, and it should work just fine.

I just looked and it appears [it's possible?](https://hub.docker.com/_/microsoft-windows-servercore-iis) to run a windows server on docker, so I could maybe test this.

**Comment by spencermountain** [comment]
yeah, so if you feel ambitious enough to poke around at `./src/worker/02-parseWiki.js` i'm sure you'd see one of the id values is messed-up, and we could add an easy condition to fix it.

otherwise, i can look at either getting a cheap windows laptop, or doing some docker-time on the weekend
cheers

**Comment by jeremythille** [comment]
OK so, change of plans today :)
I managed to get another machine on which I installed Linux Mint, and tried Dumpster-dive on it.

It works much better, so you where right when you said it was a Windows issue.

I'm still getting `====DB write error (worker 2)=== Error: key red. must not contain '.'`, so records with a `.` still get skipped, but at least the process completes.

```
👍  closing down.

     -- final count is 99,330 pages --
       took 56s
              🎉
```

I've just launched the big 65 GB English file, it should be fine.
Thanks for your help and your support :)

**Comment by spencermountain** [comment]
that's great!
will keep this open, so we can properly support windows going forward!

**Comment by jeremythille** [comment]
Following up on this, I can't manage to complete the 65 GB file, because... Mongo crashes :( (Still in Linux Mint 19.1)

The script goes fine, then suddenly (after a long time, over one or two hours), it crashes with this error :

```
#2  +500 pages  -  129ms - "List of Canadian provinces and territories by gross domestic product"
 #1  +500 pages  -  212ms - "Luigi Palma Di Cesnola"
--uncaught top-process error--
{ MongoNetworkError: connection 161 to localhost:27017 closed
    at Socket.<anonymous> (/usr/lib/node_modules/dumpster-dive/node_modules/mongodb-core/lib/connection/connection.js:352:9)
    at Object.onceWrapper (events.js:281:20)
    at Socket.emit (events.js:193:13)
    at TCP._handle.close (net.js:614:12)
  name: 'MongoNetworkError',
  errorLabels: [ 'TransientTransactionError' ],
  [Symbol(mongoErrorContextSymbol)]: {} }
{ Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:171:27)
  name: 'MongoNetworkError',
  errorLabels: [ 'TransientTransactionError' ],
  [Symbol(mongoErrorContextSymbol)]: {} }
{ Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:171:27)
  name: 'MongoNetworkError',
  errorLabels: [ 'TransientTransactionError' ],
  [Symbol(mongoErrorContextSymbol)]: {} }
/usr/lib/node_modules/dumpster-dive/node_modules/mongodb/lib/topologies/server.js:240
            throw err;
            ^

TypeError: Cannot read property 'db' of null
    at /usr/lib/node_modules/dumpster-dive/src/lib/open-db.js:22:27
    at err (/usr/lib/node_modules/dumpster-dive/node_modules/mongodb/lib/utils.js:411:14)
    at executeCallback (/usr/lib/node_modules/dumpster-dive/node_modules/mongodb/lib/utils.js:401:25)
    at err (/usr/lib/node_modules/dumpster-dive/node_modules/mongodb/lib/operations/mongo_client_ops.js:285:21)
    at connectCallback (/usr/lib/node_modules/dumpster-dive/node_modules/mongodb/lib/operations/mongo_client_ops.js:265:5)
    at server.connect (/usr/lib/node_modules/dumpster-dive/node_modules/mongodb/lib/operations/mongo_client_ops.js:353:14)
    at Server.<anonymous> (/usr/lib/node_modules/dumpster-dive/node_modules/mongodb/lib/topologies/server.js:237:11)
    at Object.onceWrapper (events.js:281:20)
    at Server.emit (events.js:193:13)
    at Pool.<anonymous> (/usr/lib/node_modules/dumpster-dive/node_modules/mongodb-core/lib/topologies/server.js:452:12)
/usr/lib/node_modules/dumpster-dive/node_modules/mongodb/lib/topologies/server.js:240
            throw err;
            ^
```
and indeed, Mongo doesn't respond on port 27017 any longer. I can't even stop or restart it. `sudo service mongod start` errors, and I must restart the whole machine. Bad luck.

First, obviously I thought it was a Mongo problem, so I searched for fixes. The version was 3.6, I upgraded it to 4.0.9, same issue. So maybe it's not a Mongo bug? I checked the logs, there's an error occurring but I can't find much on it :

```
2019-05-02T18:07:21.226+0200 I NETWORK  [conn24439] received client metadata from 127.0.0.1:48516 conn24439: { driver: { name: "nodejs", version: "3.2.3" $
2019-05-02T18:07:21.325+0200 I NETWORK  [conn24438] end connection 127.0.0.1:48512 (5 connections now open)
2019-05-02T18:07:21.408+0200 I NETWORK  [conn24439] end connection 127.0.0.1:48516 (4 connections now open)
2019-05-02T18:07:21.520+0200 E STORAGE  [conn23952] WiredTiger error (0) [1556813241:520015][13331:0x7f4e14bee700], file:collection-7-3799021168473972433.$
2019-05-02T18:07:21.520+0200 E STORAGE  [conn23952] WiredTiger error (0) [1556813241:520173][13331:0x7f4e14bee700], file:collection-7-3799021168473972433.$
[...]
2019-05-02T18:07:21.521+0200 E STORAGE  [conn23952] WiredTiger error (-31802) [1556813241:521893][13331:0x7f4e14bee700], file:collection-7-379902116847397$
2019-05-02T18:07:21.522+0200 E STORAGE  [conn23952] WiredTiger error (-31804) [1556813241:521904][13331:0x7f4e14bee700], file:collection-7-379902116847397$
2019-05-02T18:07:21.522+0200 F -        [conn23952] Fatal Assertion 50853 at src/mongo/db/storage/wiredtiger/wiredtiger_util.cpp 409
2019-05-02T18:07:21.522+0200 F -        [conn23952]

***aborting after fassert() failure
```
So maybe it's a script problem? Opening too many connections and/or spamming the DB too much? I'm not sure what to do now

**Comment by spencermountain** [comment]
yes, i've seen this before. Yikes. 
We used to try and keep a socket alive the whole time, and it would fall-over after a while for no reason, and doesn't exit cleanly. This was resolved by closing the connections once in a while.

yeah, `WiredTiger error` - these are just mongo errors. What crap.

What i'd recommend is splitting the xml file. Since you've already uncompressed the xml file, it should be [easier](https://github.com/remuslazar/node-xmlsplit) to split it into x chunks.
Pointing dumpster-dive at each file in sequence should be a clean way to do all parts of it. There are a few libraries that do this, but if it's a hassle, i'd be happy to write a script that helps split a wikipedia dump into x chunks.

if it helps, i've never seen this issue and i'm using  `v3.4.6` on ubuntu.

**Comment by jeremythille** [comment]
Splitting the big XML is a good idea. I managed to do it (30k documents per file), and I wrote a recursive script to run Dumpster on each file sequentially.

Mongo keeps crashing every now and then though. And I get new errors, like plenty of :

```
    💪  worker #7 has finished 💪 
events.js:170
      throw er; // Unhandled 'error' event
      ^

Error [ERR_IPC_CHANNEL_CLOSED]: Channel closed
    at process.target.send (internal/child_process.js:652:16)
    at p.then (/media/jer/14 TB/Wikipedia/node_modules/dumpster-dive/src/worker/index.js:78:13)
Emitted 'error' event at:
    at process.nextTick (internal/child_process.js:656:35)
    at processTicksAndRejections (internal/process/task_queues.js:79:9)

```
Then the process just hangs and I have to `killall node` and start it again.

At this point, all I'm doing is re-launching Mongo service and re-launching Dumpster again and again and again. My recursive script moves each file to a sub-folder after it's successful (which hapens sometimes), so Dumpster doesn't have to go through all of them and spend half an hour just skipping entries before resuming.
I started with 637 files, currently there are 430 to go.

**Comment by spencermountain** [comment]
hey that's great. Can you document it somewhere?

good luck!

**Comment by jeremythille** [comment]
Well, it turns out this is not a solution at all. It worked for a while, then Mongo started crashing again. I keep restarting the service, but its life expectancy is 20 - 30 seconds before it crashes again. The only thing I can do now is to manually launch and relaunch and relaunch Mongo service and Dumpster again. And there are 364 XML files to go. I had never had any stability issues with Mongo before, but with Dumpster it just keeps crashing absolutely all the time...

**Comment by spencermountain** [comment]
ah, this sounds awful. Sorry about this.
fwiw, I can't think of anything you're doing wrong. This library's been used hundreds of times on linux/mac and I don't know why it's acting up.

i looked at that Enterprise Center Tower page and don't see any funny-business. It would throw it, or log it, if there was an error. 

I'm glad you're getting closer into the details of how the library is working. It's not too complicated. If you can find a solution to improve it, please consider making a pr

don't give up!
cheers

**Comment by jeremythille** [comment]
I am very lucky because this week at work we have a MongoDB consultant. He's currently on the Linux machine trying to figure it out, hopefully everything will work in a bit :)

**Comment by spencermountain** [comment]
ahh! if he can give the library any tips, that'd be really helpful. I have no idea what i'm doing, usually.

**Comment by jeremythille** [comment]
Ok so basically he checked everything, suspected some clusters of the DB file to be corrupted so he wiped that, we relaunched the script and so far, so good. Everything is working smoothly now, fingers crossed.

He explained me why `countDocuments()` is slow, because it actually fetches all documents in the DB to return their number, unlike `count()`, which is almost instant because it only reads metadata. `countDocuments()` should only be used when the DB is sharded.

Also, we tried to increase batch_size to 500, because he said Mongo can take it easily; in fact, the more documents you give `insertMany`, the fastest (with a limit of 1000 documents at a time). Seeing no improvement in speed, we went back to the default 200.

Now it's still running smoothly without crashing :)
